Sprint 21

rodionkim63887 opened the pull request:
выполнение задания 21 спринта

rodionkim63887 committed:
Replaced @Environment viewModels with @State viewModels

rodionkim63887 committed:
Renamed ServicesFiltersViewModel to ServicesFilters as it's not a viewModel

rodionkim63887 committed:
removed unnecessary debugging print()

rodionkim63887 committed:
removed unnecessary commented lines

rodionkim63887 committed:
renamed API to NetworkService

rodionkim63887 committed:
fixed stories close button size and padding

rodionkim63887 committed:
implemented settlement and station data receiving from yandex api

rodionkim63887 committed:
implemented custom search field to counter the bug of navigationStack breaking when used with .searchable

rodionkim63887 committed:
removed unnecessary commented code

rodionkim63887 committed:
refactored the navigation and added loading state to correctly display error screens inside settings tab

rodionkim63887 committed:
changed api url as specified in yandex api documentation;
removed debugging error throwing

rodionkim63887 committed:
refactored LoadState to be usable for loading any data

rodionkim63887 committed:
fixed openAPI specification to correclty decode data

rodionkim63887 committed:
Removed unused models for mock data

rodionkim63887 committed:
activated SWIFT_STRICT_CONCURRENCY = complete

rodionkim63887 committed:
added @MainActor to OrientationObserver and Router, changed class to actor for NetworkService

rodionkim63887 committed:
removed CarriersViewModel as it is not used anymore

rodionkim63887 committed:
added ProgressView for reusability in different Views

rodionkim63887 committed:
added filter for countries to reduce amount of data processed in the app

rodionkim63887 committed:
added String extention for processing dateTime strings

rodionkim63887 committed:
renaming

rodionkim63887 committed:
deleted blank lines

rodionkim63887 committed:
refactored ServiceListView, ServiceListViewModel to work with actual data from API

rodionkim63887 committed:
Fixed unwrapping settlement and station titles for correct display

rodionkim63887 committed:
removed debugging print

rodionkim63887 committed:
replaced ProgressView in SettlementSelectionView with external component

rodionkim63887 committed:
refactored CarrierInfoView to display actual API data

rodionkim63887 committed:
refactored ServicesFilters to filter actual services

rodionkim63887 committed:
fixed/removed previews for some views

rodionkim63887 committed:
changing transfers to true should enable recieving services with transfers

NaR0RoW commented on TravelSchedule/Extentions/String+Extention.swift:
```diff
@@ -0,0 +1,30 @@
+    }
+    
+    var hourComponent: Int? {
+        let parts = self.split(separator: ":")
```
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on TravelSchedule/Router/Router.swift:
```diff
@@ -7,33 +7,63 @@
-    var path = NavigationPath()
+    var globalPath = NavigationPath()
+    var settingsPath = NavigationPath()
+    var activeTab: Int? = nil
```
`= nil` можно опустить

NaR0RoW replied:
Переменную лучше назвать `activeTabIndex`, но для TabView лучше использовать явный `enum` вместо `Int?`, чтобы избежать ошибок с индексами

NaR0RoW commented on TravelSchedule/Router/Router.swift:
```diff
@@ -7,33 +7,63 @@
 
+@MainActor
 @Observable
 final class Router: ObservableObject {
```
Лучше не совмещать атрибут `@Observable` и `ObservableObject` - это разные механики для одной работы. Нужно использовать что-то одно

NaR0RoW commented on TravelSchedule/Router/Routes/Route.swift:
```diff
@@ -7,13 +7,15 @@
-enum Route: Hashable {
-    case mainScreen
+enum GlobalRoute: Hashable {
+//    case mainScreen
```
Закомментированный код лучше удалять перед review

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/CarrierInfoView.swift:
```diff
@@ -7,6 +7,24 @@
+}
+
 struct CarrierInfoView: View {
     @Environment(\.dismiss) var dismiss
```
Поле можно сделать приватным

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/CarrierInfoView.swift:
```diff
@@ -59,8 +86,8 @@ struct CarrierInfoView: View {
                 
-                Link(destination: URL(string: "mailto:\(carrier.email)")!) {
-                    Text(carrier.email)
+                Link(destination: URL(string: "mailto:\(String(describing: carrier.email))")!) {
```
Лучше не использовать `force` операции

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/CarrierInfoView.swift:
```diff
@@ -71,23 +98,30 @@ struct CarrierInfoView: View {
                 
-                Link(destination: URL(string: "tel:\(carrier.phoneNumber)")!) {
-                    Text(carrier.phoneNumber)
+                Link(destination: URL(string: "tel:\(String(describing: carrier.phone))")!) {
```
Давай нигде не будем использовать `force` операции

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/SettlementSelectionView.swift:
```diff
@@ -7,44 +7,102 @@
+            switch viewModel.state {
+            case .loading:
+                LoadingProgressView()
+            case .success(_):
```
Лучше не использовать круглые скобки, если это явно не требуется

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/SettlementSelectionView.swift:
```diff
@@ -7,44 +7,102 @@
+                    .foregroundColor(.searchElements)
+                
+                ZStack(alignment: .leading) {
+                    if query.isEmpty {
```
Везде, где это возможно (если layout при этом не изменяется), лучше использовать `opacity` вместо `if` для сокрытия `View`. [Вот хороший пример с ссылками на `WWDC`](https://stackoverflow.com/a/57420479)

NaR0RoW commented on TravelSchedule/UI/RootView/Navigation/SettlementSelectionView.swift:
```diff
@@ -7,44 +7,102 @@
+        VStack {
+            HStack(spacing: 4) {
+                Image(systemName: "magnifyingglass")
+                    .foregroundColor(.searchElements)
```
`.foregroundColor` перестанут в будущих версиях iOS поддерживать, поэтому лучше `.foregroundStyle` использовать

NaR0RoW commented on TravelSchedule/UI/RootView/Settings/UserAgreementView.swift:
```diff
@@ -8,7 +8,7 @@
 
 struct UserAgreementView: View {
-    @StateObject private var viewModel = UserAgreementViewModel()
+    @State var viewModel: UserAgreementViewModel = UserAgreementViewModel()
```
`State property wrapper` в `SwiftUI` должен быть приватным, чтобы гарантировать, что изменение значений происходит только в теле представления

NaR0RoW commented on TravelSchedule/UI/UIComponents/LoadingProgressView.swift:
```diff
@@ -0,0 +1,17 @@
+
+import SwiftUI
+
+public struct LoadingProgressView: View {
```
Модификатор доступа `public` не нужен, так как у тебя одномодульное приложение

NaR0RoW commented on TravelSchedule/UI/ViewModels/ServiceListViewModel.swift:
```diff
@@ -7,84 +7,109 @@
+            return "\(hours) ч \(minutes) мин"
+        } else {
+            return "\(minutes) мин"
+        }
```
Лучше использовать тернарный оператор:
`return hours > 0 ? "\(hours) ч \(minutes) мин" : "\(minutes) мин"`

rodionkim63887 committed:
minor fixes based on review:
replaced .foregroundColor with .foregroundStyle;
removed @Observable from ObservableObject;
refactored Router from Int tab index to enum activeTab;
other small fixes

rodionkim63887 committed:
Added @MainActor for StoriesViewModel, marked LoadState, Route and SettingsRoute @Sendable

NaR0RoW commented:
> Если честно, я теперь еще больше запутался, что должно быть @sendable, а что нет. Когда я выставил SWIFT_STRICT_CONCURRENCY в Complete, компилятор мне никаких ворнингов не показал, хотя, кажется, обязан, если таких вещей не хватает. Единственное место, где @sendable должен пригодиться, это enum LoadState<T: Equatable & Sendable>: Equatable, Sendable, но и то я не совсем уверен в этом. Все остальные вещи, которые явно должны быть @sendable (например, сгенерированные OpenAPI модели), уже помечены автоматически =)

Ты прав, в большинстве value-типов Swift сам все сделает, но явно указывать `Sendable` на всех публичных типах - это хорошая практика, она документирует намерение, и компилятор гарантирует, что завтра при изменениях тип не станет небезопасным.
`LoadState` - это `generic enum`, лучше добавить `Sendable` явно.
Если у тебя остаются вопросы, то рекомендую обратить к наставнику, он поможет внести полную ясность
